core/src/avm1/object/bitmap_data.rs

```diff
@@ -0,0 +1,278 @@
+        let g = (self.get_green() as f64 / a).round() as u8;
+        let b = (self.get_blue() as f64 / a).round() as u8;
+
+        Color::argb(self.get_alpha(), r, g, b)
+    }
+
+    pub fn argb(alpha: u8, red: u8, green: u8, blue: u8) -> Color {
+        Color(((alpha as i32) << 24) | (red as i32) << 16 | (green as i32) << 8 | (blue as i32))
+    }
+
+    pub fn with_alpha(&self, alpha: u8) -> Color {
+        Color::argb(alpha, self.get_red(), self.get_green(), self.get_blue())
+    }
+}
+
+impl std::fmt::Display for Color {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(&format!("{:#x}", self.0))
+    }
+}
+
+impl From<Color> for i32 {
+    fn from(c: Color) -> Self {
+        c.0
+    }
+}
+
+impl From<Color> for u32 {
+    fn from(c: Color) -> Self {
+        c.0 as u32
+    }
+}
+
+impl From<i32> for Color {
+    fn from(i: i32) -> Self {
+        Color(i)
+    }
+}
+
+/// A BitmapData
+#[derive(Clone, Copy, Collect)]
+#[collect(no_drop)]
+pub struct BitmapDataObject<'gc>(GcCell<'gc, BitmapDataData<'gc>>);
+
+#[derive(Clone, Collect)]
+#[collect(no_drop)]
+pub struct BitmapDataData<'gc> {
+    /// The underlying script object.
+    base: ScriptObject<'gc>,
+
+    /// The pixels in the bitmap, stored as a array of pre-multiplied ARGB colour values
+    pixels: Vec<Color>,
+
+    width: u32,
+    height: u32,
+    transparency: bool,
+
+    disposed: bool,
+}
+
+impl fmt::Debug for BitmapDataObject<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let this = self.0.read();
+        f.debug_struct("BitmapData")
+            .field("pixels", &this.pixels)
+            .field("width", &this.width)
+            .field("height", &this.height)
+            .field("transparency", &this.transparency)
+            .finish()
+    }
+}
+
+impl<'gc> BitmapDataObject<'gc> {
+    add_field_accessors!(
+        [transparency, bool, set => set_transparency, get => get_transparency],
+        [disposed, bool, get => get_disposed],
+        [pixels, Vec<Color>, set => set_pixels],
+        [width, u32, get => get_width],
+        [height, u32, get => get_height],
+    );
+
+    pub fn empty_object(gc_context: MutationContext<'gc, '_>, proto: Option<Object<'gc>>) -> Self {
+        BitmapDataObject(GcCell::allocate(
+            gc_context,
+            BitmapDataData {
+                base: ScriptObject::object(gc_context, proto),
+                pixels: Vec::new(),
+                width: 0,
+                height: 0,
+                transparency: true,
+                disposed: false,
+            },
+        ))
+    }
+
+    pub fn get_pixels_rgba(&self) -> Vec<u8> {
+        self.0
+            .read()
+            .pixels
+            .iter()
+            .flat_map(|p| vec![p.get_red(), p.get_green(), p.get_blue(), p.get_alpha()])
+            .collect()
+    }
+
+    pub fn get_pixels(&self) -> Vec<Color> {
+        self.0.read().pixels.clone()
+    }
```

Comment:
This would be better returned as a `&[Color]` and let the caller decide whether to clone it.